ext/random/random.c

```diff
@@ -0,0 +1,1754 @@
+	zval tmp;
+	int i;
+
+	for (i = 0; i< MT_N; i++) {
+		ZVAL_LONG(&tmp, s->s[i]);
+		zend_hash_next_index_insert(data, &tmp);
+	}
+	ZVAL_LONG(&tmp, s->cnt);
+	zend_hash_next_index_insert(data, &tmp);
+	ZVAL_LONG(&tmp, s->seeded);
+	zend_hash_next_index_insert(data, &tmp);
+
+	return SUCCESS;
+}
+
+static int mersennetwister_unserialize(void *state, HashTable *data) {
+	php_random_engine_state_mersennetwister *s = (php_random_engine_state_mersennetwister *)state;
+	zval *tmp;
+	int i;
+
+	for (i = 0; i < MT_N; i++) {
+		tmp = zend_hash_index_find(data, i);
+		if (!tmp || Z_TYPE_P(tmp) != IS_LONG) {
+			return FAILURE;
+		}
+
+		s->s[i] = Z_LVAL_P(tmp);
+	}
+	tmp = zend_hash_index_find(data, MT_N); /* cnt */
+	if (!tmp || Z_TYPE_P(tmp) != IS_LONG) {
+		return FAILURE;
+	}
+	s->cnt = Z_LVAL_P(tmp);
+	tmp = zend_hash_index_find(data, MT_N + 1); /* seeded */
+	if (!tmp || Z_TYPE_P(tmp) != IS_LONG) {
+		return FAILURE;
+	}
+	s->seeded = Z_LVAL_P(tmp);
+
+	return SUCCESS;
+}
+
+static zend_object *php_random_engine_mersennetwister_new(zend_class_entry *ce) {
+	return php_random_engine_common_init(ce, &php_random_engine_algo_mersennetwister, &random_engine_mersennetwister_object_handlers);
+}
+
+const php_random_engine_algo php_random_engine_algo_mersennetwister = {
+	sizeof(uint32_t),
+	mersennetwister_dynamic_generate_size,
+	sizeof(php_random_engine_state_mersennetwister),
+	mersennetwister_generate,
+	mersennetwister_seed,
+	mersennetwister_serialize,
+	mersennetwister_unserialize
+};
+
+/* MersenneTwister end */
+
+/* Secure begin */
+
+static inline size_t secure_dynamic_generate_size(void *state) {
+	return sizeof(uint64_t);
+}
+
+static uint64_t secure_generate(void *state) {
+	uint64_t ret;
+
+	php_random_bytes_silent(&ret, sizeof(uint64_t));
+
+	return ret;
+}
+
+static zend_object *php_random_engine_secure_new(zend_class_entry *ce) {
+	return php_random_engine_common_init(ce, &php_random_engine_algo_secure, &random_engine_secure_object_handlers);
+}
+
+const php_random_engine_algo php_random_engine_algo_secure = {
+	sizeof(uint64_t),
+	secure_dynamic_generate_size,
+	0,
+	secure_generate,
+	NULL,
+	NULL,
+	NULL
+};
+
+/* Secure end */
+
+/* User begin */
+
+static inline size_t user_dynamic_generate_size(void *state) {
+	php_random_engine_state_user *s = (php_random_engine_state_user *) state;
+
+	return s->last_generate_size;
+}
+
+static uint64_t user_generate(void *state) {
+	php_random_engine_state_user *s = (php_random_engine_state_user *) state;
+	uint64_t result = 0;
+	size_t size;
+	zval retval;
+	int i;
+	char *ptr;
+
+	zend_call_known_instance_method_with_0_params(s->generate_method, s->object, &retval);
+
+	/* Store generated size in a state */
+	size = Z_STR(retval)->len;
+	s->last_generate_size = size;
+
+	if (size > 0) {
+		/* Endianness safe copy */
+		ptr = (char *) &result;
+		for (i = 0; i < size; i++) {
+			ptr[i] = Z_STR(retval)->val[i];
+		}
```

Comment:
That's right.
CI failed: https://app.travis-ci.com/github/php/php-src/jobs/560071270

I'm encountering the endian problem for the first time, and I'm trying to figure out a solution...

Reply:
The cleanest solution to fix this is by using bitshifting to construct the resulting integer:

```
result = (val[0] << 0 * 8) + (val[1] << 1 * 8) + (val[0] << 2 * 8) + (val[3] << 3 * 8);
```

By doing that you are always accessing the bytes in little endian order. The addition will happen correctly for every endianness, as that a regular addition that works with numbers.

This is what you're already doing here, just in reverse:

https://github.com/php/php-src/pull/8094/files#diff-30b7b11db85beddcf823b76a7171b28ac252a31f04e9f1a3f0985056a6f6fcbeR1280-R1282

Reply:
Hmmm, I am trying the following, but I am not getting the intended result.
```c
for (i = 0; i < size; i++) {
	result += (unsigned char) Z_STRVAL(retval)[i] << 8 * i;
}
```

Reply:
For me that loop does the right thing and correctly understands a little endian 32 bit integer:

```
final class MyEngine implements Engine {
    public function generate(): string
    {
        return \pack('V', 123);
    }
}
```

```
	if (size > 0) {
		/* Endianness safe copy */
		for (i = 0; i < size; i++) {
			result += (unsigned char) Z_STRVAL(retval)[i] << 8 * i;
		}
	} else {
		result = 0;
	}

	printf("result: %lu\n", result);
```

That outputs `result: 123` for me.

Reply:
But this will fail the next test:
```php
<?php

$native = new \Random\Engine\XorShift128Plus(1234);
$user = new class (1234) extends \Random\Engine\XorShift128Plus {};

$native_randomizer = new \Random\Randomizer($native);
$user_randomizer = new \Random\Randomizer($user);

echo 'native: ' . $native_randomizer->getInt() . PHP_EOL;
echo 'user:   ' . $user_randomizer->getInt() . PHP_EOL;

/*
Outputs: 
native: 2984280891209302271
user:   1743686824
*/
```

This only happens with Randomizer::getInt(). It may be caused by some other part. I will keep checking.

Reply:
I was confused. Thanks!